packages/core/lib/migrate.ts

```diff
@@ -89,6 +107,55 @@ const migrations: Migration[] = [
       delete updated.data.zones?.[zoneCompound];
     });
 
+    // Migrate zones created by dynamic arrays
+    const unmigratedZonesGrouped: Record<string, Record<string, Content>> = {};
+
+    Object.keys(updated.data.zones ?? {}).forEach((zoneCompound) => {
+      const [zoneId, propName] = zoneCompound.split(":");
+      const content = updated.data.zones?.[zoneCompound];
+
+      if (!content) {
+        return;
+      }
+
+      if (!unmigratedZonesGrouped[zoneId]) {
+        unmigratedZonesGrouped[zoneId] = {};
+      }
+
+      if (!unmigratedZonesGrouped[zoneId][propName]) {
+        unmigratedZonesGrouped[zoneId][propName] = content;
+      }
+    });
+
+    Object.keys(unmigratedZonesGrouped).forEach((zoneId) => {
+      updated.data = walkTree(updated.data, config, (content) => {
+        return content.map((child) => {
+          if (child.props.id !== zoneId) {
+            return child;
+          }
+
+          const migrateFn =
+            migrationOptions?.migrateDynamicZonesForComponent?.[child.type];
+
+          if (!migrateFn) {
+            return child;
+          }
+
+          const zones = unmigratedZonesGrouped[zoneId];
+          const migratedProps = migrateFn(child.props, zones);
+
+          Object.keys(zones).forEach((zoneCompound) => {
+            delete updated.data.zones?.[`${zoneId}:${zoneCompound}`];
```

Comment:
Do we need a way for the user to tell us that they migrated that zone?

I think the answers is probably not, but thought it's worth a sense-check.